Common/AR/source/Android/ArCoreWrapper.cpp

```diff
@@ -0,0 +1,933 @@
+              kNumVertices, kVerticesLeftLandscape, AR_COORDINATES_2D_TEXTURE_NORMALIZED,
+              m_transformedUvs.data());
+        break;
+      case Qt::InvertedPortraitOrientation:
+        ArFrame_transformCoordinates2d(
+              m_arSession, m_arFrame, AR_COORDINATES_2D_OPENGL_NORMALIZED_DEVICE_COORDINATES,
+              kNumVertices, kVerticesReversePortrait, AR_COORDINATES_2D_TEXTURE_NORMALIZED,
+              m_transformedUvs.data());
+        break;
+      case Qt::InvertedLandscapeOrientation:
+        ArFrame_transformCoordinates2d(
+              m_arSession, m_arFrame, AR_COORDINATES_2D_OPENGL_NORMALIZED_DEVICE_COORDINATES,
+              kNumVertices, kVerticesRightLandscape, AR_COORDINATES_2D_TEXTURE_NORMALIZED,
+              m_transformedUvs.data());
+        break;
+      default:
+        break;
+    }
+
+    m_uvsInitialized = true;
+  }
+
+  int64_t timestamp = 0;
+  ArFrame_getTimestamp(m_arSession, m_arFrame, &timestamp);
+  if (timestamp == 0)
+  {
+    // Suppress rendering if the camera did not produce the first frame yet.
+    // This is to avoid drawing possible leftover data from previous sessions if
+    // the texture is reused.
+    return;
+  }
+
+  ArFrame_acquireCamera(m_arSession, m_arFrame, &m_arCamera);
+  if (!m_arCamera)
+  {
+    emit m_arcGISArView->errorOccurred("ARCore failure", "Fails to acquire the camera.");
+    return;
+  }
+
+  // get the view and projection matrix
+  QMatrix4x4 viewMatrix;
+  QMatrix4x4 projectionMatrix;
+  ArCamera_getViewMatrix(m_arSession, m_arCamera, viewMatrix.data());
+  ArCamera_getProjectionMatrix(m_arSession, m_arCamera, 0.1f, 100.f, projectionMatrix.data());
+
+  // calculate the model-view-projection matrix. (The model matrix is the identity matrix)
+  // get the screen orientation
+  QMatrix4x4 deviceOrientationMatrix;
+  const Qt::ScreenOrientations orientation = QGuiApplication::screens().front()->orientation();
+
+  switch (orientation)
+  {
+    case Qt::PortraitOrientation:
+      // do nothing
+      break;
+
+    case Qt::LandscapeOrientation:
+      deviceOrientationMatrix.rotate(-90.0, 0.0f, 0.0f, -1.0f);
+      break;
+
+    case Qt::InvertedPortraitOrientation:
+      // not supported
+      break;
+
+    case Qt::InvertedLandscapeOrientation:
+      deviceOrientationMatrix.rotate(90.0, 0.0f, 0.0f, -1.0f);
+      break;
+
+    default:
+      // do nothing
+      break;
+  }
+
+  m_mvpMatrix = deviceOrientationMatrix * projectionMatrix * viewMatrix;
+}
+
+/*!
+  \internal
+ */
+bool ArCoreWrapper::renderVideoFeed() const
+{
+  return m_renderVideoFeed;
+}
+
+/*!
+  \internal
+ */
+void ArCoreWrapper::setRenderVideoFeed(bool renderVideoFeed)
+{
+  m_renderVideoFeed = renderVideoFeed;
+
+  // create ArCoreFrameRenderer if necessary
+  if (renderVideoFeed)
+  {
+    if (!m_arCoreFrameRenderer)
+    {
+      m_arCoreFrameRenderer.reset(new ArCoreFrameRenderer(this));
```

Comment:
No, `ArCoreFrameRenderer` is not a `QObject`. There is a argument `this` because `ArCoreFrameRenderer` must request the frame and texture parameters to the wrapper.